runtime/src/chpl-qsbr.c

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright 2004-2018 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "chplrt.h"
+#include "chpl-thread-local-storage.h"
+#include "chpl-mem.h"
+#include "chpl-atomics.h"
+#include "chpl-tasks.h"
+#include "chpl-qsbr.h"
+
+static uint64_t global_epoch = 0;
+
+struct defer_data {
+    void *data;
+    int numData;
+};
+
+// Deferred deletion list and associated epoch
+// Ideas: First thread that does not complete within a
+// reasonable amount of time gets to handle deleting it.
+// If a thread hits a checkpoint and seeing a deferred node,
+// it will process one of them; if it is unable to, it will
+// 'pass the buck' to the next problematic thread ensuring that
+// eventually it will be deleted.
+struct defer_node {
+  uint64_t targetEpoch;
+  struct defer_data ddata;
+  struct defer_node *next;
+};
+
+// Thread-local data that keeps track of our observed global state.
+struct tls_node {
+  /*
+    The current observed epoch. This is used to ensure that writers
+    do not delete an older instance that corresponds to this epoch.
+  */
+  volatile uint64_t epoch;
+  /*
+    Number of tasks that this thread has running. Required so
+    that threads without tasks (and no longer are able to call
+    checkpoints) are not taken into account.
+  */
+  volatile uint64_t nTasks;
+  /*
+    List of objects that are queued for deferred deletion. Each
+    time a writer has waited for too long on a particular thread,
+    it will 'pass the buck' for deletion to the next thread.
+  */
+  struct defer_node *deferList;
+  struct tls_node *next;
+};
+
+// List of thread-local data.
+static struct tls_node *chpl_qsbr_TLSList;
+static CHPL_TLS_DECL(struct tls_node *, chpl_qsbr_readerTLS);
+
+// Initializes TLS; should only need to be called once.
+static void init_tls(void);
+static void init_tls(void) {
+  struct tls_node *node = chpl_calloc(1, sizeof(*node));
+  node->epoch = global_epoch;
+
+  // Append to head of list
+  struct tls_node *old_head;
+  do {
+    old_head = chpl_qsbr_TLSList;
+    node->next = old_head;
+  } while (!atomic_compare_exchange_weak_uintptr_t((uintptr_t *) &chpl_qsbr_TLSList, (uintptr_t) old_head, (uintptr_t) node));
```

Comment:
We like things to fit in 80 columns. Didn't measure this one but it looks really long.